### DISCUSSION TITLE
Keys do not work when changing layout

### DISCUSSION THREAD

**Discussion by Kushiko** [davatorium/rofi#1898] [(Possible) Issues]
Hello, when typing in English everything is fine. As soon as I change the layout, no keyboard keys work except letters (Ecs, backspace does not work). What configs should I attach and how to solve the problem?

**Comment by DaveDavenport** [comment]
Start with some obvious basic information?
* Version of rofi
* What layout you are switching too
* Steps to reproduce
* etc.

**Comment by Kushiko** [comment]
rofi 1.7.3
to rus
I switch back to eng

**Reply by DaveDavenport** [reply]
Neither `rus` or `eng` are valid layouts on my system.   I have no idea how to reproduce using 'I switch back to eng'.

So I came up with a setup myself:
```bash
setxkbmap  -layout us,ru  -option grp:win_space_toggle
```
and I can switch between layouts without issue and everything keeps running and text input as well as escape/backspace keeps working.

**Reply by Kushiko** [reply]
Yes exactly

**Reply by DaveDavenport** [reply]
What distribution are you running? (I am on ubuntu 20.04 lts).
Because it works here for me with both latest rofi as version 1.7.3

**Reply by Kushiko** [reply]
debian testing trixie

**Reply by DaveDavenport** [reply]
works fine here on debian testing.

**Reply by Kushiko** [reply]
On the English keyboard, everything works great for me. Apparently rofi does not understand the Russian layout. If I start typing in Russian, I have to re-login to rofi, this is quite inconvenient... I use the filebrowser in rofi, and I have to search for some files in Russian, which doesn’t work(

**Reply by DaveDavenport** [reply]
As mentioned before, with the info you gave me, it works perfectly fine here. 
I can enter rofi in english enter text, switch to ru layout, enter Cyrillic script text , switch back use special keys etc.
I can both search for directories with english or russian names.

Rofi does not understand 'layouts' this is all handled by libxkbcommon. 

What do you mean with 'have to re-login to rofi' ?

**Reply by Kushiko** [reply]
Perhaps I made a mistake in my explanation. I don't re-login to rofi. Let's say I need to erase a character when I type in Russian. To do this, I need to change the layout to Eng, erase the symbol and change the layout back. This is not very convenient.. By 're-login' I meant restart. Just Esc and start again. Maybe I should change something in libxkbcommon?

**Reply by DaveDavenport** [reply]
Given I cannot reproduce the issue and I am not to familiar with multi-layout input, I am afraid I don't know.
For me here everything seems to be working fine, I just tested backspace too, and this works.

Just to be sure? you are on X11 and not wayland?

**Reply by Kushiko** [reply]
Yes, X11. I3wm

**Reply by Kushiko** [reply]
Do you have any more ideas?

**Reply by DaveDavenport** [reply]
Given I cannot reproduce the issue given the information you gave me and I am not to familiar with multi-layout input, I am afraid I don't know.
For me here everything seems to be working fine and this works.
I even made a similar setup with debian and did everything to test with the info I got.

**Reply by wwalker** [reply]
@ILYFKA What exactly are you using to set up layout swapping?  Like:
```
setxkbmap -model pc104 -layout ru,us -option grp:win_space_toggle
```
Or your config if you are doing layout switching setup via config files like:
```
/etc/X11/xorg.conf.d/00-keyboard.conf
Section "InputClass"
        Identifier "system-keyboard"
        MatchIsKeyboard "on"
        Option "XkbLayout" "cz,us"
        Option "XkbModel" "pc104"
        Option "XkbVariant" ",dvorak"
        Option "XkbOptions" "grp:win_space_toggle"
EndSection
```

**Reply by Kushiko** [reply]
What if I switch the layout using Alt+Shift? And why is cz (Czech) specified in the config?

**Reply by DaveDavenport** [reply]
>  And why is cz (Czech) specified in the config?

Because it is an illustrative example? the code snippets are used to clarifying the question asked.

**Reply by Kushiko** [reply]
I put the ru layout first. Now rofi works in Russian, but does not work in English. Apparently he only sees the main layout. We need a way to combine the esc and backspace buttons on different layouts. They do have different codes, I looked it up through the xev tool

**Reply by DaveDavenport** [reply]
They don't here?  For me escape has the same keycode on both layouts.
I think we are still missing information.  Can you **please** give us the complete configuration of your keyboard setup.

As a work around you can set both keycodes in the keybinding.

@sardemff7  Can we detect layout changes and reload the key mapping?

**Reply by wwalker** [reply]
@ILYFKA Please run the following and post the results:
```
$ setxkbmap -print -verbose 10
```

**Reply by wwalker** [reply]
Also, please specify the specific keyboard you are using and what the keyboard's default layout is.

**Reply by Kushiko** [reply]
setxkbmap -print -verbose 10: 

```
Setting verbose level to 10
locale is C
Trying to load rules file ./rules/evdev...
Trying to load rules file /usr/share/X11/xkb/rules/evdev...
Success.
Applied rules from evdev:
rules:      evdev
model:      pc105
layout:     us,ru
variant:    qwerty
options:    grp:alt_shift_toggle
Trying to build keymap using the following components:
keycodes:   evdev+aliases(qwerty)
types:      complete
compat:     complete
symbols:    pc+us+ru:2+inet(evdev)+group(alt_shift_toggle)
geometry:   pc(pc105)
xkb_keymap {
        xkb_keycodes  { include "evdev+aliases(qwerty)" };
        xkb_types     { include "complete"      };
        xkb_compat    { include "complete"      };
        xkb_symbols   { include "pc+us+ru:2+inet(evdev)+group(alt_shift_toggle)"        };
        xkb_geometry  { include "pc(pc105)"     };
};
``` 
My keyboard: Bloody B188

**Reply by Kushiko** [reply]
Anyone else have any ideas? This error is bothering me(

**Reply by DaveDavenport** [reply]
Did you try the workaround I suggested?

**Reply by Kushiko** [reply]
which one?

**Reply by wwalker** [reply]
@ILYFKA :

https://github.com/davatorium/rofi/discussions/1898#discussioncomment-7002259

"As a work around you can set both keycodes in the keybinding."

**Reply by wwalker** [reply]
Like @DaveDavenport , I cannot reproduce your problem.  On my keyboards backspace is the same in ru or us...

**Reply by Kushiko** [reply]
It's also possible that it's my keyboard, but I'm not sure, I don't have a second keyboard to check. I don't quite understand the keycodes substitution. I read about this on the forum, but didn't get anywhere...

**Comment by lufte** [comment]
I'm also suffering from this issue. I'm on Debian as well, XFCE, and it breaks when I switch from `English (US)` to `English (US, intl., with dead keys)`. I can provide any extra detail you need from my setup.

**Reply by DaveDavenport** [reply]
I did not manage to reproduce this one, so more information would be welcome. 
Can you give some details about versions? (rofi, debian, xfce, ) and 

```bash
setxkbmap -print -verbose 10
```

**Reply by lufte** [reply]
Sure thing. I'm on Debian testing, running rofi 1.7.3, XFCE 4.18. I'm using the XFCE keyboard settings to switch between layouts with Right Ctrl + Right Shift. My exact behavior is that, when I'm on US international, I can only type letters. Enter, Escape, arrows, backspace and Tab don't work.

This didn't always happen, but I didn't realize until today (when I read this discussion) that the cause was the keyboard layout, so I didn't fully investigate the issue because I thought it happened randomly. What I do remember though, is that I checked my dpkg logs around the time I noticed this was happening and rofi hadn't been updated in a while, so I suspected it was caused by a dependency.

Lastly, here's the output you needed:
```
Setting verbose level to 10
locale is C
Trying to load rules file ./rules/evdev...
Trying to load rules file /usr/share/X11/xkb/rules/evdev...
Success.
Applied rules from evdev:
rules:      evdev
model:      pc101
layout:     us,us
variant:    ,intl
options:    grp:rctrl_rshift_toggle,ctrl:nocaps
Trying to build keymap using the following components:
keycodes:   evdev+aliases(qwerty)
types:      complete
compat:     complete
symbols:    pc+us+us(intl):2+inet(evdev)+ctrl(nocaps)+group(rctrl_rshift_toggle):1+group(rctrl_rshift_toggle):2
geometry:   pc(pc101)
xkb_keymap {
        xkb_keycodes  { include "evdev+aliases(qwerty)" };
        xkb_types     { include "complete"      };
        xkb_compat    { include "complete"      };
        xkb_symbols   { include "pc+us+us(intl):2+inet(evdev)+ctrl(nocaps)+group(rctrl_rshift_toggle):1+group(rctrl_rshift_toggle):2"   };
        xkb_geometry  { include "pc(pc101)"     };
};
```

**Reply by DaveDavenport** [reply]
thanks, I cannot reproduce it on i3 on debian testing . I will install xfce later to try

**Reply by DaveDavenport** [reply]
Yeah, I can reproduce the issue in XFCE.
Weird that exactly the same setup works fine under gnome/i3.

**Reply by DaveDavenport** [reply]
`~~It works for me under xfce 4.16, but not 4.18..  What did xfce change. ~~`

see below

**Reply by DaveDavenport** [reply]
I did some tests..  It works fine in 1.7.5  it breaks in 1.7.3..

Can you please update rofi and try again?

**Reply by lufte** [reply]
Confirmed! 1.7.5 has not landed on Debian yet, but I just built it from source and it works for both keyboard layouts. Many thanks!

**Reply by wwalker** [reply]
May need to contact whoever is responsible for building/maintaining rofi in debian, 1.7.4 released 14 months ago...

Fedora already has 1.7.5.

Thank you to whoever maintains rofi on Fedora!!

And thank you to Qball and the other rofi devs!

**Reply by lufte** [reply]
According to https://packages.debian.org/sid/rofi, @jpleau is the Debian maintainer of rofi. Let's see if he is active on Github, otherwise I'll try emailing him.

Jason sorry to invoke you here, we are wondering if it's possible to upgrade rofi to 1.7.5 to fix this bug related to XFCE.

**Comment by Kushiko** [comment]
I updated to Rofi 1.7.5 and everything worked!
